[DSCP-281] Getting rid of implicit network identities

volhovm opened the pull request:
My goal is to get rid of predefined public-ip-based identities, replacing them with random identities, which are supposed to be communicated when the connection is initiated.

I'll squash everything later.

Heimdell commented on code/network/lib/Loot/Network/ZMQ/Common.hs:
```diff
@@ -87,82 +59,84 @@ ztLog Logging{..} level t = do
-        inet_ntoa (hostAddress ent)
+    splitBy :: Eq a => a -> [a] -> [[a]]
+    splitBy _ [] = []
+    splitBy d s' = x : splitBy d (drop 1 y) where (x,y) = L.span (/= d) s'
```
[splitOn](http://hackage.haskell.org/package/split-0.2.3.3/docs/Data-List-Split.html#v:splitOn)

volhovm replied:
Nice function, though it in some external library so I obviously won't import a lib for just a oneliner.

Heimdell commented on code/network/lib/Loot/Network/ZMQ/Server.hs:
```diff
@@ -173,7 +171,11 @@ runBroker = do
 
     let frontToListener = \case
-            (cId:"":msgT:msg) -> do
+            [cId,"","getid"] -> do
```
```suggestion
            [cId, "", "getid"] -> do
```
We aren't running low on screen width and neither we're going to print the code on narrower analogs of paper towels. Please, do not forget spaces.

volhovm replied:
> Please, do not forget spaces.

I don't put them intentionally because I find it easier to read.

>We aren't running low on screen width

I am actually limited to 82 symbols width.

Heimdell commented on code/network/lib/Loot/Network/ZMQ/Server.hs:
```diff
@@ -173,7 +171,11 @@ runBroker = do
+                Z.sendMulti ztServFront $
+                  NE.fromList [cId,"",unZTInternalId ztOurNodeId]
+                ztLog ztServLogging Debug "Received request connection, replied with our id"
+            (cId:"":"m":msgT:msg) -> do
```
```suggestion
            cId : "" : "m" : msgT : msg -> do
```

Heimdell commented on code/network/lib/Loot/Network/ZMQ/Server.hs:
```diff
@@ -173,7 +171,11 @@ runBroker = do
+            (cId:"":"m":msgT:msg) -> do
                 ztEnv <- atomically $ runMaybeT $ do
                     lId <- MaybeT $ Map.lookup (MsgType msgT) <$> readTVar ztMsgTypes
                     MaybeT $ Map.lookup lId <$> readTVar ztListeners
```
I can understand what is going on here, if I concentrate on those words between `[run]MaybeT`, but its hard. Can you help, please?

volhovm replied:
Yes sure. 
L180: get the listener id that uses this message type.
L181: resolve these listener id into the listener

MaybeT will fail if any of these will fail.

volhovm commented on code/network/lib/Loot/Network/Class.hs:
```diff
@@ -80,19 +80,23 @@ import Loot.Network.BiTQueue (BiTQueue (..))
-    { _uprAdd :: Set nId -- ^ Peers to add/connect.
-    , _uprDel :: Set nId -- ^ Peers to delete/disconnect.
+    { _uprAdd :: [nId] -- ^ Peers to add/connect.
+    , _uprDel :: [nId] -- ^ Peers to delete/disconnect.
```
No, I've just simplified an interface. Semantics stays the same -- these fields will be `nub`bed before anything else.

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -40,8 +41,43 @@ import Loot.Log.Internal (Level (..), Logging (..), NameSelector (..), logNameSe
+data PeersInfoVar = PeersInfoVar
+    { _pivConnected  :: TVar (HashMap ZTNodeId ZTInternalId)
+    -- ^ Peers we're connected to.
+    , _pivConnecting :: TVar (HashMap ZTNodeId (Integer, Z.Socket Z.Dealer, SocketAdapter))
```
I'll wrap it.

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -40,8 +41,43 @@ import Loot.Log.Internal (Level (..), Logging (..), NameSelector (..), logNameSe
+peersRevMap :: PeersInfoVar -> STM (HashMap ZTInternalId ZTNodeId)
+peersRevMap piv = do
+    l <- map swap . HMap.toList <$> readTVar (piv ^. pivConnected)
+    let res = HMap.fromList l
```
Negative, since we don't use it anywhere else.

Heimdell replied:
For clarity, dude.

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -80,32 +116,33 @@ updateHeartbeat states nodeId = do
+                          if hbs ^. hbLiveness == 1
+                          then (True, (nodeId, hbs & hbInactive .~ True))
+                          else (False, (nodeId, hbs & hbLiveness -~ 1
+                                                    & hbNextPoll .~
```
I don't mind padding-dependent indentation with lenses.

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -123,15 +160,43 @@ instance T.Show InternalRequest where
-    (add',del')
+-- Get hashmap keys as a hashset.
+hmKeys :: HashMap k a -> HashSet k
+hmKeys = HSet.fromMap . fmap (const ())
```
I've replaced `fmap` with `HMap.map`

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -123,15 +160,43 @@ instance T.Show InternalRequest where
+    -- Normalising the request
+    let both = L.nub $ _uprDel `L.intersect` _uprAdd
+    let add' = L.nub _uprAdd L.\\ both
+    let del' = L.nub _uprDel L.\\ both
```
I don't see how that would follow. `where` case is not created to merely decouple the function logic, it also does more (it can loop, it can be called several times). Comments are just comments.

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -177,20 +241,19 @@ data ZTNetCliEnv = ZTNetCliEnv
-createNetCliEnv (ZTGlobalEnv ctx ztLogging) peers = liftIO $ do
-    ztCliBack <- Z.socket ctx Z.Router
+createNetCliEnv :: MonadIO m => ZTGlobalEnv -> [ZTNodeId] -> m ZTNetCliEnv
+createNetCliEnv globalEnv@ZTGlobalEnv{..} peers = liftIO $ do
```
Our style guide suggests no space before/after the `@`, see https://github.com/serokell/serokell-util/blob/master/serokell-style.md#function-declaration, `putValueInState` example.

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -177,20 +241,19 @@ data ZTNetCliEnv = ZTNetCliEnv
+                            <*> newTVarIO HMap.empty
     ztHeartbeatInfo <- newTVarIO mempty -- initialise it later
     ztSubscriptions <- newTVarIO mempty
     ztMsgTypes <- newTVarIO mempty
```
> This function really needs to be torn apart to smaller functions.

No it doesn't. Why?

>Aside of that, i assume that ztHeartbeatInfo should be initialised later and rest of the horde may be left uninitialised.

`ztHeartbeatInfo` is actually already initialised (with an empty map), I've removed the comment.

> Does the lack of ztHeartbeatInfo initialisation (later) breaks the code?

It's just a mapping `nodeid -> heartbeatstate`, so we put related data into this map when we update peers.

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -212,31 +275,98 @@ termNetCliEnv ZTNetCliEnv{..} = liftIO $ do
     Z.close ztCliBack
     adapterRelease ztCliSubAdapter
     Z.close ztCliSub
+    peersRes <- atomically $ HMap.elems <$> readTVar (ztPeers ^. pivConnecting)
```
I guess it's pretty obvious what's going on...

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -212,31 +275,98 @@ termNetCliEnv ZTNetCliEnv{..} = liftIO $ do
+        -- Modify actual variables content.
+        modifyTVar (ztPeers ^. pivConnected) $! \p ->
+            p `HMap.difference` HMap.fromList (map (,()) toDel)
+        let addInfo = map (\(s,a) -> (curTime,s,a)) $ sockets `zip` adapters
```
It's an intermediate computation which doesn't need any external datatype. If I used a datatype whenever I needed a pair that'd be disastrous. I'll change to `zipWith`, thanks.

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -212,31 +275,98 @@ termNetCliEnv ZTNetCliEnv{..} = liftIO $ do
+
+        -- Modify actual variables content.
+        modifyTVar (ztPeers ^. pivConnected) $! \p ->
+            p `HMap.difference` HMap.fromList (map (,()) toDel)
```
Why would I convert a list to hashset to then use `fromSet`, if `HMap.fromList . map (,())` is several times shorter? What's a use?

volhovm replied:
>would be easy to understand

It is easy to understand.

>immediately obvious to reader

It is immediately obvious to the reader. 

I'll rewrite it if you provide more evidence that this looks complex (other's opinions).

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -212,31 +275,98 @@ termNetCliEnv ZTNetCliEnv{..} = liftIO $ do
+        modifyTVar (ztPeers ^. pivConnected) $! \p ->
+            p `HMap.difference` HMap.fromList (map (,()) toDel)
+        let addInfo = map (\(s,a) -> (curTime,s,a)) $ sockets `zip` adapters
+        modifyTVar (ztPeers ^. pivConnecting) $! \p ->
```
You mean the `$!` operator? It's probably a leftover from some debugging session, I'll remove it thx.

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -212,31 +275,98 @@ termNetCliEnv ZTNetCliEnv{..} = liftIO $ do
         modifyTVar ztHeartbeatInfo $
-            foldr (.) id (map (\nId -> at nId .~ Nothing) (Set.toList toDel)) .
-            foldr (.) id (map (\nId -> at nId .~ Just initHbs) (Set.toList toAdd))
+            foldr (.) id (map (\nId -> at nId .~ Nothing) toDel)
```
This is a pattern you should generally recognise: for each `e \in toDel` i generate a "deletion action" and then combine them with `(.)`. That's just how you apply several modifications at once :man_shrugging:

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -212,31 +275,98 @@ termNetCliEnv ZTNetCliEnv{..} = liftIO $ do
+            "changePeers: preconnecting (sending req) to " <> show toConnect
+
+    -- Release sockes that were not used.
+    let dataLeft = drop (length toConnect) $ sockets `zip` adapters
```
Well, yes, I pre-generate a list of DEALER sockets, take the amount I need for the initialisation, and then release the tail.

>Is this socket usage mechanism represented as actual piece of code somewhere

Yes, you can find almost all the interactions with the DEALER socket by grepping "dealer" in this module and in the server module.

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -212,31 +275,98 @@ termNetCliEnv ZTNetCliEnv{..} = liftIO $ do
+-- Continues connection request (moves peer from "connecting" to
+-- "connected" state).
+contConnectionReq :: ZTGlobalEnv -> ZTNetCliEnv -> ZTNodeId -> ZTInternalId -> IO ()
+contConnectionReq ZTGlobalEnv{..} ZTNetCliEnv{..} nodeId iId = do
```
It's "continue" in the original sense of the word, I'm not meaning it's related to `Cont` in any sense (of course).

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -212,31 +275,98 @@ termNetCliEnv ZTNetCliEnv{..} = liftIO $ do
+    curTime <- getCurrentTimeMS
+    resourcesM <- atomically $ do
+        connected <- HMap.toList <$> readTVar (ztPeers ^. pivConnected)
+        let clash = isJust $ find (\(_,iId') -> iId' == iId) connected
```
Fixed.

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -212,31 +275,98 @@ termNetCliEnv ZTNetCliEnv{..} = liftIO $ do
+            (_,sock,adapter) <-
+                fromMaybe (error $ "contConnectionReq: cant find by nodeId") .
+                HMap.lookup nodeId <$>
+                readTVar (ztPeers ^. pivConnecting)
```
>Eye is tricked

That's a default style used for about 3 years already. Please start a thread in #code-style

> mutable stm-hasmap

Number of dependencies/complexity tradeoff. I might introduce them later for the sake of speeding something up.

volhovm commented:
@Heimdell I've pushed some of the changes you asked me for, I've replied to most of the comments, but (I admit and I'm sorry) I deleted about 5 small comments regarding spaces, because I found them unrelated and I want reviewers to follow less conversations. I've also squashed all the previous commits, you can check the last one (it contains fixes only). Also please review the logic.

Heimdell commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -330,62 +509,86 @@ runBroker = do
+
+                -- Choosing at random if not provided (both are nothing).
+                (iIdFinal :: Maybe ZTInternalId) <-
+                    maybe (fmap snd <$> choosePeer) (pure . Just) iIdM
```
There is hell of a lot `let`-functions in code in `runBroker`. Can you make them `where`-functions instead, so the body will find itself near the function header?

volhovm replied:
I can't because they use the contexts which I `view` in the beginning of `runBroker`.

Heimdell replied:
I recommend duplicating the reading statement to them.

I cannot fit this function on my mind at once.

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -201,7 +264,7 @@ createNetCliEnv (ZTGlobalEnv ctx ztLogging) peers = liftIO $ do
     let ztCliLogging = ztLogging & logNameSelL %~ modGivenName
     let ztCliEnv = ZTNetCliEnv {..}
```
Yes, this is a default record initialisation pattern, recall `ApplicativeDo` and parsers.

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -212,31 +275,98 @@ termNetCliEnv ZTNetCliEnv{..} = liftIO $ do
+
+            -- 2 seconds heuristical delay before heartbeating worker
+            -- starts noticing these nodes.
+            let initHbs = HBState hbIntervalMin hbLivenessMax (curTime + 2000) False
```
I've just removed it.

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -259,6 +389,40 @@ reconnectPeers ZTNetCliEnv{..} nIds = liftIO $ do
+            prev <- readTVar (piv ^. pivConnecting)
+            let (stay,leave) = L.partition predicate (HMap.toList prev)
+            writeTVar (piv ^. pivConnecting) (HMap.fromList stay)
+            pure (leave,map (view $ _2 . _2) stay)
```
Yeah but `view $ _2 . _2` is much more straightforward than what you suggest :man_shrugging:

Heimdell replied:
> straightforward

Our target is to write _manageable_ code that everyone of us will be able to support. "Straightforwardness" should go after that.

volhovm replied:
Ok, fixed.

Martoon-00 commented on code/network/lib/Loot/Network/ZMQ.hs:
```diff
@@ -30,12 +28,14 @@ Client:
-
+* Every time we need to connect to some server, we open a DEALER socket,
+  ask server's ROUTER for his zmq identity, disconnect and destroy this
+  DEALER and use ROUTER <-> ROUTER communication.
```
Btw could you explain why so? I was wondering yet when I read the whole code.

volhovm replied:
In distributed ZMQ networks you basically have two options: ROUTER <-> ROUTER and DEALER <-> ROUTER. You should understand that the second option means that the client has a DEALER for each peer it's connected to. This is more or less what we're doing at the "connection handshake" step, right. Both approaches have advantages and disadvantages. The most straightforward advantage of DEALER <-> ROUTER is that we don't care about identifiers as whenever we get the message from peer P we receive it on the socket related to P, so the whole idea of identifiers can be ditched. The issue is that in our networking prototype we also have PUB <-> SUB and we really want updates to have node identity inside (so thus we can query the block with hash H from N when we get update H from the node N). So in order to solve this problem in case we use DEALER <-> ROUTER we can either use identities (but then why would we switch to DEALER <-> ROUTER) or introduce a SUB socket per peer -- which I find somewhat counterintuitive as all the PUB/SUB manuals recommend using one SUB for many PUBs.

volhovm commented on code/network/lib/Loot/Network/ZMQ/Server.hs:
```diff
@@ -211,8 +213,13 @@ runBroker = do
                   SBListener _lId msg -> processMsg msg
                   SBFront             -> whileM (canReceive ztServFront) $
                                          Z.receiveMulti ztServFront >>= frontToListener
```
`frontToListener` does the `\case` itself, it's much more readable this way.

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -212,31 +279,100 @@ termNetCliEnv ZTNetCliEnv{..} = liftIO $ do
+        -- message.
+        forM_ (toConnect `zip` sockets) $ \(z,d) -> do
+            Z.connect d $ ztNodeIdRouter z
+            Z.sendMulti d $ NE.fromList ["","getid"]
```
How can I make it as a pattern and a function alias at the same time? e.g. I now have `pattern GetId` which I can match on, but I can't use it in this particular `zendMulti d ["", GetId]`. If there's a solution I'll refactor this.

volhovm replied:
Update: I have done it straightforwardly without pattern synonyms.

Martoon-00 replied:
Pattern synonyms can be symmetric, thus I thought you could use them as value alias at the same time. Something like
```
pattern ConnectionReq cId t = [cId,"",t]
pattern Msg cId t msgT msg = cId:"":t:msgT:msg
```

Martoon-00 replied:
@volhovm I'd like you to respond on this (ack or nack), then I approve.

> Update: I have done it straightforwardly without pattern synonyms.
I'm actually more concerned about overall message format being duplicated, not only tags.

volhovm replied:
I think that somehow replacing this will only obscure reader's understanding of how ZMQ sockets work. Yes it looks kind of unusual and strange, but if you are acquainted with ZMQ itself it becomes obvious and even (imho) more straightforward rather than using any kind of wrappers. So I'd like to leave it as it is. Though your suggestion about internal messages was good :)

Martoon-00 replied:
Okay, thank you :+1:

volhovm commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -212,31 +279,100 @@ termNetCliEnv ZTNetCliEnv{..} = liftIO $ do
+            -- 2 seconds heuristical delay before heartbeating worker
+            -- starts noticing these nodes.
+            let initHbs = HeartbeatState hbIntervalMin hbLivenessMax (curTime + 2000) False
+            modifyTVar ztHeartbeatInfo $ at nodeId .~ Just initHbs
```
Nice catch. Yes, technically the following combination is possible (1-2 can commute).
1. HB worker says "peer X is off, broker pls reconnect"
2. User asks to disconnect from X.
3. Broker disconnects X.
4. Broker processes the reconnect request from the HB worker and fails. 

I guess I will just modify `reconnectPeers` in the way that is first runs STM where it checks whether peers were not disconnected, runs HB update (and thus can't panic), and then after STM it does `disconnect >> connect`.

Martoon-00 replied:
I will review this tomorrow, my brain is too off now to check those +18/-10

volhovm commented:
@Martoon-00 I've addressed all your comments, please re-review.

Heimdell commented on code/network/lib/Loot/Network/ZMQ/Client.hs:
```diff
@@ -80,32 +116,33 @@ updateHeartbeat states nodeId = do
+    threadDelay 50000
+    curTime <- getCurrentTimeMS
+    let modMap :: Map ZTNodeId HBState -> (Set ZTNodeId, Map ZTNodeId HBState)
+        modMap hbMap =
```
Wat. It was called `decideReconnect` before?

volhovm replied:
No, it wasn't, I've renamed it. You can read all the comments you've left: I either replied with text or put a thumb up (for "fixed") or down (for "useless"). You can also see the separate commit in the commit history that addresses the changes you asked to.

volhovm commented:
>Ok, I get it, you don't want to make code nice. 

I've fixed more than a half of things you asked for given that more than 3/4 of them in total were obviously subjective.

>
